api/cluster.go

```diff
@@ -505,3 +508,27 @@ func (s *Server) peerQuerySpeculativeChan(ctx context.Context, data cluster.Trac
 
 	return resultChan, errorChan
 }
+
+func (s *Server) indexMetaTagRecordUpsert(ctx *middleware.Context, req models.IndexMetaTagRecordUpsert) {
+	if s.MetricIndex == nil {
+		response.Write(ctx, response.NewMsgp(200, &models.MetaTagRecordUpsertResult{}))
+		return
+	}
+
+	record := idx.MetaTagRecord{
+		MetaTags: req.MetaTags,
+		Queries:  req.Queries,
+	}
+
+	result, created, err := s.MetricIndex.MetaTagRecordUpsert(req.OrgId, record)
+	if err != nil {
+		response.Write(ctx, response.NewError(http.StatusBadRequest, err.Error()))
```

Comment:
err can be non-nil for 3 reasons:
1) invalid input
2) tag support is disabled
3) "MetaTagRecordUpsert: Unable to find a slot to insert record"

for case 1, StatusBadRequest is appropriate. for case 2 too I think. but not for 3!
look through some of the other code, I think there's a few places where we need to differentiate user error vs internal error so we can return the proper code.

Reply:
Good point. Not sure what's the best way to solve that right now. This comes back to your suggestion that input validation should happen earlier, so bad requests don't even reach the index. 
I think for now I'll just make `upsert()` return two error values, one for bad request errors and one for internal errors. 
Once this PR is merged I'll make another one that moves the input validation of all tag-related stuff into the api request handling, then we can remove the second error return value again.

Reply:
You can do it by decorating the error. We have an error type for this. We use this in other places for the same reason

Reply:
Actually, I think it would be better to continue returning only one error from those functions, but define the error message somewhere in the package scope. Then, in the package scope, there could be a list of errors that are server-side errors, all others are bad request errors. That way there's no need to clutter the function signatures.

Reply:
Regarding your case 2), should this be considered a request error? I would have thought this should be considered an internal error (resulting in `5xx`) if that feature is disabled on the server.

Reply:
Done

Reply:
missed a spot, i added it. also not sure why you changed case 2 to be an internal error, that doesn't make sense to me. i switched it back to badrequest like it was before.